pkg/controller/drivelvgcontroller/drivelvgcontroller.go

```diff
@@ -0,0 +1,378 @@
+	resourceName := req.Name
+	ctx, cancelFn := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancelFn()
+
+	log := d.log.WithFields(logrus.Fields{"method": "Reconcile", "name": resourceName})
+
+	drive := &drivecrd.Drive{}
+	if err := d.client.ReadCR(ctx, resourceName, "", drive); err == nil {
+		return d.reconcileDrive(ctx, drive)
+	}
+	lvg := &lvgcrd.LogicalVolumeGroup{}
+	if err := d.client.ReadCR(ctx, resourceName, "", lvg); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Warnf("Failed to read LVG and Drive CRs, resource with name %s is not found", resourceName)
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	return d.reconcileLVG(lvg)
+}
+
+// reconcileLVG perform logic for LVG reconciliation
+func (d *Controller) reconcileLVG(lvg *lvgcrd.LogicalVolumeGroup) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "reconcileLVG",
+	})
+	var (
+		status = lvg.Spec.GetStatus()
+		health = lvg.Spec.GetHealth()
+		name   = lvg.GetName()
+	)
+	// If LVG status is failed or Health is not good, try to reset its AC size to 0
+	if status == apiV1.Failed || health != apiV1.HealthGood {
+		return ctrl.Result{}, d.resetACSizeOfLVG(name)
+	}
+	// If LVG is already presented on a machine but doesn't have AC, try to create its AC using annotation with
+	// VG free space
+	size, err := getFreeSpaceFromLVGAnnotation(lvg.Annotations)
+	if err != nil {
+		log.Errorf("Failed to get free space from LVG %v, err: %v", lvg, err)
+		if err == errTypes.ErrorNotFound {
+			err = nil
+		}
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, d.createLVGACIfNotExistsOrUpdate(lvg, size)
+}
+
+// reconcileDrive preforms logic for drive reconciliation
+func (d *Controller) reconcileDrive(ctx context.Context, drive *drivecrd.Drive) (ctrl.Result, error) {
+	var (
+		health = drive.Spec.GetHealth()
+		status = drive.Spec.GetStatus()
+	)
+	switch {
+	case health != apiV1.HealthGood || status != apiV1.DriveStatusOnline:
+		return d.handleInaccessibleDrive(ctx, drive.Spec)
+	default:
+		return d.createACIfNotExistOrUpdate(ctx, drive.Spec)
+	}
+}
+
+// createACIfNotExistOrUpdate tries to create AC for drive or update its size if AC already exists
+func (d *Controller) createACIfNotExistOrUpdate(ctx context.Context, drive api.Drive) (ctrl.Result, error) {
+	log := d.log.WithFields(logrus.Fields{
+		"method": "createACIfNotExistOrUpdate",
+	})
+	driveUUID := drive.GetUUID()
+	size := drive.GetSize()
+	// if drive is not clean, size is 0
+	if !drive.GetIsClean() {
+		size = 0
+	}
+	ac, err := d.cachedCrHelper.GetACByLocation(driveUUID)
+	switch {
+	case err == nil:
+		lvg, err := d.crHelper.GetLVGByDrive(ctx, driveUUID)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// If LVG exists for Drive, we delete drive AC, because CSI uses LVG AC
+		if lvg != nil {
+			if err := d.client.DeleteCR(context.WithValue(ctx, base.RequestUUID, ac.Name), ac); err != nil {
```

Comment:
We don't delete AC anymore. Set size to 0 instead.

Reply:
This code doesn't relate to health change of Drive. This functionality was added to handle situation, when drive LVG is presented in cluster and drive AC exists. We should delete it as CSI uses LVG AC in this case. But I suppose this is not really probable situation, so this code can be removed.

Reply:
Removed this code